### DISCUSSION TITLE
`web_sys::js_sys::Function`'s different behaviors

### DISCUSSION THREAD

**Discussion by alivarastepour** [wasm-bindgen/wasm-bindgen#3895] [Q&A]
I have been working with `web_sys::js_sys::Function` in 2 different scenarios; first one is inside a struct as follows:
```rust
    #[derive(Serialize, Deserialize)]
    #[wasm_bindgen]
    pub struct MyStruct {
        my_fn: Function,
    }
```
The second one is as a parameter in a function; like this:
```rust
    pub fn some_function(
        my_fn: &Function,
    ) -> ();
```
I can use the later without any problems, which is, I can call it from my JS code and pass a JS function as an argument to it and let `some_function` call that argument. But in the former, I've been facing a issue related to the `web_sys::js_sys::Function` type not implementing the `Deserialize<'_>` trait. 

My question is, what exactly is the technical issue with (de)serializing functions?(`web_sys::js_sys::Function`, to be more precise). Why can't we employ the same behavior(as in later scenario, which is passing only a reference to the JS function I presume) for structs in case of functions as well? Is There a workaround to use this type in such structs anyways?

**Comment by Liamolucko** [comment]
> My question is, what exactly is the technical issue with (de)serializing functions?(`web_sys::js_sys::Function`, to be more precise). Why can't we employ the same behavior(as in later scenario, which is passing only a reference to the JS function I presume) for structs in case of functions as well? Is There a workaround to use this type in such structs anyways?

Serde is mostly designed for serialising/deserialising data from formats like JSON, YAML, and TOML, and in those contexts deserialising to a JavaScript function makes no sense. However, if you're using `serde-wasm-bindgen`, it has a feature for passing through JavaScript values that you can use: https://docs.rs/serde-wasm-bindgen/latest/serde_wasm_bindgen/#preserving-javascript-values.

**Reply by alivarastepour** [reply]
Yeah this actually works, but raises 2 questions for me; first of, what happens to the data which is not serialized? I mean, if we can preserve values, why serialize them in the first step? second, how is a preserved field interpreted(by ABI) in the end?
